GitTfs.VsCommon/TfsHelper.Common.cs

```diff
@@ -296,23 +296,70 @@ private void GetRootChangesetForBranch(IList<RootBranch> rootBranches, string tf
+                    // In this case, the code previously assumed C3 is the root changeset and would only check for merge history in it.
+                    // Now, the code does not assume any given changeset is the branch root and instead crawls its history in batches to find the first changeset
+                    //  with merge history and assumes that changeset is the root.
+
+                    const int batchSize = 100;
+
+                    IEnumerable<MergeInfo> branchChangesetsInTargetBranch = null;
+                    for (var batchNumber = 1; branchChangesetsInTargetBranch == null; batchNumber++)
                     {
-                        throw new GitTfsException("An unexpected error occurred when trying to find the root changeset.\nFailed to find first changeset for " + tfsPathBranchToCreate);
+                        var changesetsToRetrieve = batchNumber*batchSize;
+
+                        var changesetEnumerable = VersionControl.QueryHistory(tfsPathBranchToCreate, VersionSpec.Latest, 0, RecursionType.Full,
+                            null, null, null, changesetsToRetrieve, false, false, false, true).Cast<Changeset>();
+
+                        if (batchNumber > 1)
+                            changesetEnumerable = changesetEnumerable.Skip((batchNumber - 1)*batchSize).Take(batchSize);
+
+                        // ToListed because inspecting the enumerable during debugging was resulting in TFS timeouts
+                        var changesets = changesetEnumerable.ToList();
+
+                        // If our batch has no results, there's nothing left to check. We're done.
+                        if (!changesets.Any())
+                            break;
+
+                        for (var index = 0; index < changesets.Count; index++)
+                        {
+                            var changeset = changesets.ElementAt(index);
+                            var branchChangesetsInTargetBranchForBatch = GetMergeInfo(tfsPathBranchToCreate, tfsPathParentBranch, changeset.ChangesetId, lastChangesetIdToCheck);
+
+                            if (branchChangesetsInTargetBranchForBatch.Any())
+                            {
+                                branchChangesetsInTargetBranch = branchChangesetsInTargetBranchForBatch;
+                                break;
+                            }
+                        }
                     }
 
-                    var mergedItemsToFirstChangesetInBranchToCreate = GetMergeInfo(tfsPathBranchToCreate, tfsPathParentBranch, firstChangesetInBranchToCreate.ChangesetId, lastChangesetIdToCheck);
+                    if (branchChangesetsInTargetBranch == null)
+                    {
+                        throw new GitTfsException("Failed to find a branch root changeset for " + tfsPathBranchToCreate);
+                    }
 
                     string renameFromBranch;
-                    var rootChangesetInParentBranch =
```

Comment:
I don't have a strong advice on that naming and don't really know which is the best. Keep it that way if you feel that's better...